libs/cardano-data/src/Data/UMap.hs

```diff
@@ -0,0 +1,691 @@
+-- eval (dom rewards' ◁ iRTreasury (_irwd ds) :: RewardAccounts (Crypto era))
+
+domRestrict :: (Ord cr, Ord ptr) => VMap coin cr pool ptr k v -> Map k u -> Map k u
+domRestrict (Rewards (UnifiedMap tmap _)) m = intersectDomPLeft p m tmap
+  where
+    p _ (Triple (SJust _) _ _) = True
+    p _ _ = False
+domRestrict (Delegations (UnifiedMap tmap _)) m = intersectDomPLeft p m tmap
+  where
+    p _ (Triple _ _ (SJust _)) = True
+    p _ _ = False
+domRestrict (Ptrs (UnifiedMap _ pmap)) m = Map.intersection m pmap
+
+-- ==========================
+
+type Tbor x = (Typeable x, ToCBOR x)
+
+instance
+  (Tbor coin, Tbor ptr, Ord ptr, ToCBOR pool) =>
+  ToCBOR (Trip coin pool ptr)
+  where
+  toCBOR (Triple coin ptr pool) =
+    encodeListLen 3 <> toCBOR coin <> toCBOR ptr <> toCBOR pool
+
+instance
+  (FromCBOR coin, Ord ptr, FromCBOR ptr, FromCBOR pool) =>
+  FromSharedCBOR (Trip coin pool ptr)
+  where
+  type Share (Trip coin pool ptr) = Interns pool
+  fromSharedCBOR is =
+    decodeRecordNamed "Triple" (const 3) $
+      do
+        a <- fromCBOR
+        b <- fromCBOR
+        c <- fromShareCBORfunctor is
+        pure (Triple a b c)
+
+instance
+  (Tbor coin, Tbor ptr, Tbor cred, ToCBOR pool, Ord ptr) =>
+  ToCBOR (UMap coin cred pool ptr)
+  where
+  toCBOR (UnifiedMap tripmap ptrmap) =
+    encodeListLen 2 <> encodeMap toCBOR toCBOR tripmap <> encodeMap toCBOR toCBOR ptrmap
+
+-- One can do this using fromSharedPlusCBOR and Lenses and lift, but I think this is much clearer
+-- using fromSharedPlusCBOR'. By giving a name to the shared state, i.e. (a,b) we can use
+-- individual pieces of it where we need, without involving Lense combinators or lift. The only
+-- drawback is we that we must build the bigger state  (internsFromMap tripmap <> a, b) by hand.
+
+instance
+  (Ord cred, FromCBOR cred, Ord ptr, FromCBOR ptr, FromCBOR coin, FromCBOR pool) =>
+  FromSharedCBOR (UMap coin cred pool ptr)
+  where
+  type
+    Share (UMap coin cred pool ptr) =
+      (Interns cred, Interns pool)
+  fromSharedPlusCBOR' (a, b) =
+    decodeRecordNamed "UnifiedMap" (const 2) $ do
+      tripmap <- decodeMap (interns a <$> fromCBOR) (fromSharedCBOR b)
+      ptrmap <- decodeMap fromCBOR (interns a <$> fromCBOR)
+      pure (UnifiedMap tripmap ptrmap, (internsFromMap tripmap <> a, b))
+
+-- =================================================================
+
+data Tag coin cred pool ptr k v where
+  Rew :: Tag coin cred pool ptr cred coin
+  Del :: Tag coin cred pool ptr cred pool
+  Ptr :: Tag coin cred pool ptr ptr cred
+
+class View coin cred pool ptr k v where
+  tag :: Tag coin cred pool ptr k v
+
+-- ===================================================
+-- derived operations
+
+findWithDefault :: (Ord cred, Ord ptr) => a -> k -> VMap coin cred pool ptr k a -> a
+findWithDefault a1 k vm =
+  case lookup k vm of
+    Just a2 -> a2
+    Nothing -> a1
+
+size :: VMap coin cred pool ptr k a -> Int
+size x = foldl' (\count _v -> count + 1) 0 x
```

Comment:
`Map.size` is O(1), but this implementation is O(n). I think manual pattern match on underlying maps and using `Map.size`  would be a good optimization

Reply:
The problem is that there is no constant time size for KeyMap, so you will aways resort to expensive algorithms.  This approach avoids tricky interations at the boundary of IntMap and KeyMap

Reply:
I think you might have mixed up two PRs, in particular mine with utxo+SplitMap and this one. `UMap` doesn't use `KeyMap`, instead it relies on `Map` which hash O(1) size op

Reply:
You are right, I got these mixed up. But, there is another subtle thing we need to consider. A VMap is a view of a UnifiedMap, Which is a pair of maps, So we would need something like
`size (Delegations (Unified tripmap ptrmap)) = ...
 size (Rewards (Unified tripmap ptrmap)) = ...
 size (Ptrs (Unified tripmap ptrmap)) = ... `
The size of the view for Delegations and Rewards depends upon the shape of the Triple in the tripmap, NOT the size of tripmap. So the best we can do is inspect each of the triples for their shape, and count the ones with the right shape. That inspecting the Triples is exactly what the foldl' does.  So we can do a bit better, because the Ptr view size can be computed directly from the 'ptrmap'.  So here is a better, but not optimal solution.
`size :: VMap coin cred pool ptr k a -> Int
size (Ptrs (UnifiedMap _ ptrmap)) = Map.size ptrmap
size x = foldl' (\count _v -> count + 1) 0 x
`